docs/long_term_roadmap.md

```diff
@@ -0,0 +1,242 @@
+duplication and irregularity across the backends.
+
+Moving forward, we would like for the backends to share more code and for their
+op support to be more aligned with each other. Since the backend contract today
+includes "all" of PyTorch's operators, it is very costly to duplicate the
+lowering of so many ops across backends. Additionally, there are 3
+forward-looking efforts that intersect with this effort:
+
+- [StableHLO](https://github.com/openxla/stablehlo) - this is a dialect
+  initially forked from MHLO which intends to create a stable support surface
+  area for what today is our "at head" dependency on MHLO. MHLO is a fairly
+  complete op set, so it is very attractive to have "almost all" models
+  bottleneck through a stable interface like StableHLO. StableHLO is currently
+  under relatively early development, but already delivers on many of the goals
+  of stability.
+- [TCP](https://github.com/llvm/torch-mlir/issues/1366) - this is a dialect
+  which could serve a role very similar to MHLO, while providing community
+  ownership. TCP is still in early planning phases, but there is strong
+  alignment with the StableHLO effort. One byproduct of TCP that is expected to
+  be very valuable is to incorporate the robust dynamic shape strategy from
+  Linalg into an MHLO-like dialect, and there is a strong desire from StableHLO
+  developers to adopt this once proven in TCP.
+-
+  [PrimTorch](https://dev-discuss.pytorch.org/t/tracing-with-primitives-update-0/577) -
+  is an effort on the PyTorch side to decompose PyTorch operators into a
+  smaller set of primitive ops. This effort could effectively reduce the op
+  surface area at the Torch-MLIR level a lot, which would make the duplication
+  across backends less of anissue. But it still leaves open a lot of
+  questions, such as how to control decompositions.
+
+This is overall less important than the frontend refactor, because it is "just
+more work" for us as Torch-MLIR developers to support things in the current
+infrastructure, while the frontend refactor directly affects the user
+experience.
+
+As the above efforts progress, we will need to make decisions about how to adopt
+the various technologies. The main goal is consolidating the bottleneck point
+where the O(100s-1000s) of ops in PyTorch are reduced to a more tractable O(100)
+ops. There are two main ways to accomplish this:
+
+- Future A: We concentrate the bottleneck step in the "Backend contract ->
+  StableHLO/MHLO/TCP" lowering path. This gets us a stable output for most
+  things. The cascaded/transitive lowerings then let us do O(100) lowerings from
+  then on down. (exact details are not worked out yet, and depend on e.g. TCP
+  adoption, etc.)
+- Future B: PrimTorch concentrates the bottleneck step on the PyTorch side.
+
+These two efforts synergize, but the need for cascaded lowerings is much less if
+PrimTorch solves the decomposition problem on the PyTorch side.
```

Comment:
So, what are the main concerns for us solving this problem on the PyTorch side? Or is it just the lack of coverage/maturity of this project?

Reply:
I think it is coverage/maturity, together with the right kinds of supported, documented user-facing (user = toolchain implementor here) controls for decompositions. @ngimel mentioned that such tools for controlling decompositions were planned (or already existing?), so I think we are all aligned here.

Reply:
Original program semantics lost in decomposition should be considered harmful. 

Early milestone of a compiler project is usually focused on operator coverage, so decomposing is good in terms of supporting more models: hey, at least I only need to cover the primitives, and I let framework decomposes the rest. However, one does not simply make a compiler (and hardware) just to support more models. Users want better performance, otherwise they could just use plain CPU (or CUDA w/o fancy stuff like fp16 tensor core or transformer engine). With more hard-wired accelerators built into these chips, or even with handcrafted custom op kernels, I often find myself stuck in a weird situation: a large number of local patterns need to be matched to recover those computation semantics (some w/ decent perf improvement opportunities) decomposed in upper level of the stack.

Alternatively, I find myself lazy (or incompetent) in writing all these pattern matches, and I simply save the effort by ditch a particular set of decompositions. With a specific model and backend, I could figure out the optimal decomposition configuration. But with more models, more hardware, and an evolving compiler stack? The complexity only multiplies.

No one likes such a hardware lottery, but facing the end of Moore's law, there is no other way out.